### ISSUE TITLE
Auth Error continued, with nginx/docker implementations.

### ISSUE THREAD

**Issue by digitalextremist** [espocrm/espocrm#2726]
The problem is, once you log out, you cannot log back in until you clear the cache of your browser.

That right there seems to be a clue of some kind. If I delete the cookies but do not clear the cache... error persists. If I leave the cookies and clear the cache, error gone.

---

This continues the following other references:
- https://forum.espocrm.com/forum/installation-and-upgrade-help/82794-v7-2-login-popup-issue?p=91042#post91042
- https://github.com/espocrm/espocrm/issues/2422#issuecomment-1532233995

@TobiasDeBruijn has a clue to this in the link above. I think? I am trying to draw that out to fix either the documentation for how to correctly setup `nginx`, the `docker` provisioning approach, `espocrm` core fix, or patch to one of the libraries we depend on.

It seems to die here:

https://github.com/espocrm/espocrm/blob/9056ad493b193cfbce48c2bb085571fb9504aca7/application/Espo/Core/Authentication/Authentication.php#L130

That said, this issue could still be a duplicate after all my research, still not finding the answer already there. But I have searched high and low for a solution to this one. I can barely describe the problem but I can reproduce it.

---

I am using the `docker` implementation, with `docker-compose.yml` provisioning, and an `nginx` reverse-proxy providing `HTTPS` and `WSS` access in to the container from the world, with virtual hosts versus port addresses, etc.

---

The manifestation of the error is just a nondescript `Auth Error` like this:

![image](https://user-images.githubusercontent.com/21095/235813299-0d171ad8-1b01-4bd4-acca-c87e6e46ca02.png)

And then the `Log in` button stays disabled. So you reload.

> **Now there is also no login screen.** Just `Auth Error`

If you clear the cookies but not the cache, you get the login screen back, but actually logging in draws `Auth Error` and the loop starts again. Clear cookies for the FDQN. Get login screen back. Enter valid username and password. `Auth Error`. Refresh. No login screen. Clear cookies... infinite repeat.

---

@yurikuzn popped in right away, saying it's `nginx` not passing `Espo-Authorization` or ( some part of `HTTP Authorization` ) ... but I dropped in his test, and it does come through:

![image](https://user-images.githubusercontent.com/21095/235814022-65855f7e-0c24-4ec0-b2cf-079762fa8b7f.png)

---

I notice that value is the same value already in the `Authorization` header, but in that instance it is prefaced with `Basic` and then one whitespace character. I tried rewriting the `Authorization` header as an `Espo-Authorization` header, and it had no effect. Partly because it was not being blocked probably, and the `Espo-Authorization` header is sort of not the issue at all?

It seems like `auth-token` is set only, `auth-token-secret` is not set, and even with the cookie is removed, until you clear the cache, it cannot write the `auth-token-secret` for some reason, seemingly involving `nginx`. But this doesn't happen on a fresh browser!

If you buy a new computer, load up `Linux Mint` and surf to your `EspoCRM` login... or use a `Private Window` / `Incognito` ... you can login the first time, no problem. Logout though? Never getting back in there.

---

This will never come back with a value equal to the `authToken->getSecret()` value we already know.

---

To be super technical, _technically_ it seems to fail [here](https://github.com/espocrm/espocrm/blob/937b2adf572181b740af40f1370f7251265ff529/application/Espo/Core/Authentication/Authentication.php#L194-L196):

```
if ($result->isFail()) {
    return $this->processFail($result, $data, $request);
}
```

But that is because of the lack of data, going back to the real point of convergence between a correct login/refresh of the UI. Right there, a functioning or clean browser will get a value back from `$request->getCookieParam(self::COOKIE_AUTH_TOKEN_SECRET);` and then get all the way to:

https://github.com/espocrm/espocrm/blob/937b2adf572181b740af40f1370f7251265ff529/application/Espo/Core/Authentication/Authentication.php#L264-L266

---

Like I said, this is posted so that whatever the solution is can be documented and the implementations corrected for everyone. I see a lot of people with this issue, and I have not solved it in my instances, though some others seem to have fixed it?

@TobiasDeBruijn your example of bringing all the cookie keys into one hash again, in one `Set-Cookie` I presume, does not seem to carry over to the `docker` / `nginx` implementation of `EspoCRM`? Or does it?

Pardon the confusion here, but there are several strange wrinkes:

- It does seem to die in a dependency of `EspoCRM` -- not in `EspoCRM` itself?
- It seems to go away if you clear the cache, but not if you click the cookies only?

Would be awesome to be able to logout, then log back in. Important functionality I guess I took for granted until this issue.

**Comment by digitalextremist** [comment]
Last thing. The `Auth log` is filling up with these:

![image](https://user-images.githubusercontent.com/21095/235817085-51477410-12b6-48e0-9d27-e37b9b931bca.png)

You will see when I clear cache, I can login again, with one successful login in there.

**Comment by digitalextremist** [comment]
Tracing further:

When we hit `PostDestroyAuthToken` at logout, that token is the one the failing login is pulling up. It is pulling up an inactive token, and not seeming to get a fresh token at login. Tracing...

**Comment by digitalextremist** [comment]
Continuing to look for differences:

Successful logins first hit: `/api/v1/App/user`

Compared to unsuccessful which first hit: `/api/v1/`

**Comment by digitalextremist** [comment]
This seems to be heading away from cookies and over to cache:

Why would one login hit `/api/v1` with an `Espo-Authorization` that looks like this:

```
ZG9ub3Zhbjo0YWE4MWFjMmMyNzY0MTE1NzdhYWRiYWVlYmRlYTVlOQ== 
```

And a successful login hits `/api/v1/App/user` with an `Espo-Authorization` header that looks like this:

```
ZG9ub3ZhbjpoMzFNdVIyVnlZ
```

**Comment by digitalextremist** [comment]
Interesting...

If you clear cache ( or buy a new computer and assemble it and open a browser fresh ) and login...

... then delete your cookie ...

You lose your session on the frontend. But when you login, it goes through.

Only after using `Logout` hits `/api/v1/App/destroyAuthToken` can we no longer login.

**Comment by digitalextremist** [comment]
Browser console is also reporting a `401` on `/api/v1` when there is no login form. When the cookies are removed, that is what goes away, to allow your next failed attempt to login.

**Comment by digitalextremist** [comment]
> ```
> ZG9ub3Zhbjo0YWE4MWFjMmMyNzY0MTE1NzdhYWRiYWVlYmRlYTVlOQ== 
> ```

Is apparently `Base64.encode(userName  + ':' + password);` which _does_ get set as both `Authorization` and `Espo-Authorization` at first.

**Comment by digitalextremist** [comment]
And `inactive` tokens piling up:

![image](https://user-images.githubusercontent.com/21095/235831394-91c39001-0c78-4e48-bd88-22dd58cdc8dc.png)

**Comment by arkadywtf** [comment]
Could you please share your docker-compose.yml file? I have successfully installed EspoCRM with docker on nginx and I did not encounter the issue with logging out and in.

**Comment by yurikuzn** [comment]
> Why would one login hit `/api/v1` with an `Espo-Authorization` that looks like this:
> 
> ```
> ZG9ub3Zhbjo0YWE4MWFjMmMyNzY0MTE1NzdhYWRiYWVlYmRlYTVlOQ== 
> ```
> 
> And a successful login hits `/api/v1/App/user` with an `Espo-Authorization` header that looks like this:
> 
> ```
> ZG9ub3ZhbjpoMzFNdVIyVnlZ
> ```

When the front-end logs in (on login form), it send a request to the `/api/v1/App/user` endpoint with a username:password in the authorization header. It returns a generate auth token and generated auth token secret (in cookies header).

All subsequent requests are sending username:auth-token (instead of the password). Auth token secret is also sent (implicitly by the browser).

Could be cookies being stripped.

**Comment by digitalextremist** [comment]
Added `proxy_pass_header Set-Cookie;` and still get `Auth Error` until clearing cache. 

Only cache, not even cookies. I only do this, to get back in:

![image](https://user-images.githubusercontent.com/21095/236042907-133ee00f-ceda-43c8-9b4b-199692bfbc9e.png)

The cookie is still there, when going back to login. But the cache is removed. What could be in the cache that would be preventing login?

**Comment by digitalextremist** [comment]
@arkadywtf would you mind giving an example of a working configuration if it differs from the docs?

I wonder if you are using the `fpm` image for example? Very eager to chase down why this is failing.

From there, I think you are one of the people to talk to about additional extensions, and look forward to that once this is fixed.

**Comment by RabeeBrahimi** [comment]
have a look at this https://github.com/tmachyshyn/espocrm-environment/blob/master/production/php8.0-nginx-mysql/docker-compose.yml
there are other example of production for php 7.4 etc

**Comment by arkadywtf** [comment]
Will post my docker-compose.yml later today and yes I'm using `fpm-alpine` image for nginx setup.

**Comment by arkadywtf** [comment]
# EspoCRM Docker Compose

```yml
---
version: '3.8'

################################################################################
# VOLUMES
###############################################################################
volumes:
  mysql-data:
    name: espocrm-mysql-data

  espocrm-data:
    name: espocrm-app-data

################################################################################
# NETWORKS
################################################################################
networks:
  app-network:
    name: espocrm-network
    driver: bridge

################################################################################
# SERVICES
################################################################################
services:
  # ------------------------------------------------------------
  # MySQL Database
  # ------------------------------------------------------------
  mysql:
    image: mariadb:10.6
    container_name: mysql
    hostname: mysql
    restart: unless-stopped

    networks:
      - app-network

    environment:
      MYSQL_ROOT_PASSWORD: root_password
      MYSQL_DATABASE: espocrm
      MYSQL_USER: espocrm
      MYSQL_PASSWORD: database_password

    volumes:
      - mysql-data:/var/lib/mysql:rw

  # ------------------------------------------------------------
  # EspoCRM
  # ------------------------------------------------------------
  espocrm:
    image: espocrm/espocrm:7-fpm-alpine
    container_name: espocrm
    restart: unless-stopped

    networks:
      - app-network

    environment:
      ESPOCRM_DATABASE_HOST: mysql
      ESPOCRM_DATABASE_USER: espocrm
      ESPOCRM_DATABASE_PASSWORD: database_password
      ESPOCRM_ADMIN_USERNAME: admin
      ESPOCRM_ADMIN_PASSWORD: password
      ESPOCRM_SITE_URL: "http://localhost:8080"

    volumes:
      - espocrm-data:/var/www/html:rw
```

# Nginx Docker Compose

```yml
---
version: '3.8'

################################################################################
# VOLUMES
################################################################################
volumes:
  web-data:
    name: espocrm-app-data
    external: true

################################################################################
# NETWORKS
################################################################################
networks:
  app-network:
    name: espocrm-network
    external: true

################################################################################
# SERVICES
################################################################################
services:
  # ------------------------------------------------------------
  # Web Server
  # ------------------------------------------------------------
  nginx:
    image: nginx:stable-alpine
    restart: unless-stopped

    ports:
      - 8080:80

    networks:
      - app-network

    volumes:
      - web-data:/var/www/html:ro
      - ./espo.conf:/etc/nginx/conf.d/default.conf:ro
```

# Nginx Config File

```cfg
server {
    listen 80 default_server;
    listen [::]:80 default_server;

    server_name localhost; # domain name

    charset utf-8;
    index index.html index.php;

    client_max_body_size 50M;

    keepalive_timeout 300;
    types_hash_max_size 2048;

    server_tokens off;
    fastcgi_send_timeout 300;
    fastcgi_read_timeout 300;

    gzip on;
    gzip_types text/plain text/css text/javascript application/javascript application/json;
    gzip_min_length 1000;
    gzip_comp_level 9;

    root /var/www/html/public; # path to public dir

    location /client {
        root /var/www/html; # path to espocrm root dir
        autoindex off;

        location ~* ^.+.(js|css|png|jpg|svg|svgz|jpeg|gif|ico|tpl)$ {
            access_log off;
            expires max;
        }
    }

    location = /favicon.ico { access_log off; log_not_found off; }
    location = /robots.txt  { access_log off; log_not_found off; }

    location ~ \.php$ {
        fastcgi_pass espocrm:9000;
        include fastcgi_params;
        fastcgi_index index.php;
        fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
        fastcgi_param QUERY_STRING $query_string;
    }

    location /api/v1/ {
        if (!-e $request_filename){
            rewrite ^/api/v1/(.*)$ /api/v1/index.php last; break;
        }
    }

    location /portal/ {
        try_files $uri $uri/ /portal/index.php?$query_string;
    }

    location /api/v1/portal-access {
        if (!-e $request_filename){
            rewrite ^/api/v1/(.*)$ /api/v1/portal-access/index.php last; break;
        }
    }

    location ~ /(\.htaccess|\web.config|\.git) {
        deny all;
    }
}
```

**Comment by digitalextremist** [comment]
Thank you both so much @arkadywtf and @RabeeBrahimi!

Between those two, I am very confident I will be able to get my instances going properly. Will report back.

**Comment by digitalextremist** [comment]
@yurikuzn curious: Why is the `docker` implementation not using `fpm` like in the examples above? I am following the docker `README` there, and it is very different:

https://hub.docker.com/r/espocrm/espocrm

**Comment by digitalextremist** [comment]
Also, @arkadywtf does this implementation support `WebSockets`?

@yurikuzn mentioned something here about `WebSockets`:

https://forum.espocrm.com/forum/extensions/92340-communications-pack

In the docker hub instance, it has a websocket container. Curious if the `fpm` implementation without that: only having two versus three containers, also covers websockets?

**Comment by digitalextremist** [comment]
Final thing for @yurikuzn and @arkadywtf, pardon all the questions on implementing properly:

All the examples use a secondary port such as `8080` then do not have a configuration to proxy down to `443` to use the FQDN directly, without having to list a port in the URL. I feel like the issue I have happens most likely there, in getting it down to `443` which would involve another proxy, in all examples. How is everyone getting from `8080` to `443` without `Auth Error`?

**Comment by digitalextremist** [comment]
AH HA!

Without changing away from the official `docker-compose.yml` example, I thought differently about this:

- [x] Others are using `fpm` and presumably reverse proxies, and not having `Auth Error` -- why me?
- [x] Removing cache fixes the issue, with cookies not being cleared at all -- why cache?

Then I looked at a successful login and saw the `Response` headers had a `36h` expiry.

And I noticed that `/api/v1/App/user` was pulled as a JSON body with that `36h` expiry...

And sure enough, my reverse proxy configuration was adding a `36h` cache command in `nginx` configuration.

Removed that `36h` expiration and ta-da... no more `AuthError` after I cleared cache one last time to get rid of the `36h` expiry on `/api/v1/App/user` in the `docker-compose.yml` in the documentation, and posted by @tmachyshyn to `hub.docker` 

Thank you so much everyone for all help on this one. The issue was my local configuration, in my reverse proxy. I noticed in that last question above that no one had reverse proxy configurations showing... and that was where the problem happened.
